google/resource_compute_attached_disk.go

```diff
@@ -0,0 +1,234 @@
+package google
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	compute "google.golang.org/api/compute/v1"
+)
+
+func resourceComputeAttachedDisk() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAttachedDiskCreate,
+		Read:   resourceAttachedDiskRead,
+		Delete: resourceAttachedDiskDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: resourceAttachedDiskImport,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(300 * time.Second),
+			Delete: schema.DefaultTimeout(300 * time.Second),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"disk": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: compareSelfLinkOrResourceName,
+			},
+			"instance": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: compareSelfLinkOrResourceName,
+			},
+			"project": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"zone": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAttachedDiskCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	attachedInstance := d.Get("instance").(string)
+	zone, err := getZoneForAttachedDisk(d, config, attachedInstance)
+	if err != nil {
+		return err
+	}
+
+	instanceName := GetResourceNameFromSelfLink(attachedInstance)
+	diskName := GetResourceNameFromSelfLink(d.Get("disk").(string))
+
+	attachedDisk := compute.AttachedDisk{
+		DeviceName: diskName,
+		Source:     fmt.Sprintf("/projects/%s/zones/%s/disks/%s", project, zone, diskName),
+	}
+
+	op, err := config.clientCompute.Instances.AttachDisk(project, zone, instanceName, &attachedDisk).Do()
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s", instanceName, diskName))
+
+	waitErr := computeSharedOperationWaitTime(config.clientCompute, op, project,
+		int(d.Timeout(schema.TimeoutCreate).Minutes()), "disk to attach")
+	if waitErr != nil {
+		d.SetId("")
+		return waitErr
+	}
+
+	return resourceAttachedDiskRead(d, meta)
+}
+
+func resourceAttachedDiskRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+	d.Set("project", project)
```

Comment:
Ah yes. I was parsing the `zone` but treating it pretty different from project. I could either add a `getProjectFromSelfLink` like I did for `zone` or just allow the user to specify / override if they choose.

I'm leaning towards parsing the project link just like I did with zone because if a user specified a different project/zone than what was set on the compute instance it could cause strange behavior. If you think that's over thinking things I'm also happy to just add them both as optional.

Reply:
Total drive by comment, but: why not set the zone and project from the instance we `Get` from the API? That way users don't need to repeat themselves when the values exist in the self link of the instance, and we still find the correct values.

Reply:
I think the thing that's making this weird is if the user is only referencing the `name` when running `Create`. In which case we have to look for the values specified more globally. Otherwise I could just grab all the data from the self link. TBH... we don't even really need to set the zone/project in state at all now that I moved to using full/relative links to the compute instance instead of just `name`. I'll throw up a quick commit and see if it's an improvement.

Reply:
Oh yup - there should always be a zone and project to make the GET to at Read time set in both Create and Update!